examples/DGmethods/ex_001_dycoms.jl

```diff
@@ -58,100 +57,363 @@ eprint = {https://doi.org/10.1175/MWR2930.1}
+  state.ρu    = SVector(ρ*u, ρ*v, ρ*w) 
   state.ρe    = E
   state.moisture.ρq_tot = ρ * q_tot
-end   
+end
+
+
+# TODO: temporary; move to new CLIMA module
+# TODO: add an option to reduce communication: compute averages
+# locally only
+function gather_diagnostics(dg, Q)
+  mpirank = MPI.Comm_rank(MPI.COMM_WORLD)
+  nranks = MPI.Comm_size(MPI.COMM_WORLD)
+
+  grid = dg.grid
+  topology = grid.topology
+  N = polynomialorder(grid)
+  Nq = N + 1
+  Nqk = dimensionality(grid) == 2 ? 1 : Nq
+  nrealelem = length(topology.realelems)
+  nvertelem = topology.stacksize
+  nhorzelem = div(nrealelem, nvertelem)
+
+  nstate = 6
+  nthermo = 5
+  host_array = Array ∈ typeof(Q).parameters
+  localQ = host_array ? Q.realdata : Array(Q.realdata)
+  thermoQ = zeros(Nq * Nq * Nqk, nthermo, nrealelem)
+  vgeo = grid.vgeo
+  localvgeo = host_array ? vgeo : Array(vgeo)
+
+  for e in 1:nrealelem  
+    for i in 1:Nq*Nq*Nqk
+      z = localvgeo[i,grid.x3id,e]
+
+      rho_node  = localQ[i,1,e]
+      u_node    = localQ[i,2,e] / localQ[i,1,e]
+      v_node    = localQ[i,3,e] / localQ[i,1,e]
+      w_node    = localQ[i,4,e] / localQ[i,1,e]
+      etot_node = localQ[i,5,e]/localQ[i,1,e]
+      qt_node   = localQ[i,6,e]/localQ[i,1,e]
+
+      e_int = etot_node - 1//2 * (u_node^2 + v_node^2 + w_node^2) - grav * z
+
+      ts = PhaseEquil(e_int, qt_node, rho_node)
+      Phpart = PhasePartition(ts)
+      thermoQ[i,1,e] = Phpart.liq
+      thermoQ[i,2,e] = Phpart.ice
+      thermoQ[i,3,e] = qt_node-Phpart.liq-Phpart.ice
+      thermoQ[i,4,e] = ts.T
+      thermoQ[i,5,e] = liquid_ice_pottemp(ts)
+    end
+  end
+
+  fluctT = zeros(Nq * Nq * Nqk, nthermo, nrealelem)    
+  fluctQ = zeros(Nq * Nq * Nqk, nstate, nrealelem)
+  varQ   = zeros(Nq * Nq * Nqk, nstate, nrealelem)
+
+  rho_localtot = sum(localQ[:, 1, :])
+  U_localtot   = sum(localQ[:, 2, :])
+  V_localtot   = sum(localQ[:, 3, :])
+  W_localtot   = sum(localQ[:, 4, :])
+  e_localtot   = sum(localQ[:, 5, :])
+  qt_localtot  = sum(localQ[:, 6, :])
+
+  qliq_localtot  = sum(thermoQ[:, 1, :])
+  qice_localtot  = sum(thermoQ[:, 2, :])
+  qvap_localtot  = sum(thermoQ[:, 3, :])
+  T_localtot     = sum(thermoQ[:, 4, :])
+  theta_localtot = sum(thermoQ[:, 5, :])
+
+  rho_tot   = MPI.Reduce(rho_localtot, +, 0, MPI.COMM_WORLD)
+  U_tot     = MPI.Reduce(U_localtot, +, 0, MPI.COMM_WORLD)
+  V_tot     = MPI.Reduce(V_localtot, +, 0, MPI.COMM_WORLD)
+  W_tot     = MPI.Reduce(W_localtot, +, 0, MPI.COMM_WORLD)
+  e_tot     = MPI.Reduce(e_localtot, +, 0, MPI.COMM_WORLD)
+  qt_tot    = MPI.Reduce(qt_localtot, +, 0, MPI.COMM_WORLD)
+  qliq_tot  = MPI.Reduce(qliq_localtot, +, 0, MPI.COMM_WORLD)
+  qice_tot  = MPI.Reduce(qice_localtot, +, 0, MPI.COMM_WORLD)
+  qvap_tot  = MPI.Reduce(qvap_localtot, +, 0, MPI.COMM_WORLD)
+  T_tot     = MPI.Reduce(T_localtot, +, 0, MPI.COMM_WORLD)
+  theta_tot = MPI.Reduce(theta_localtot, +, 0, MPI.COMM_WORLD)
+
+  if mpirank == 0
+    rho_avg   = rho_tot / (size(localQ, 1) * size(localQ, 3) * nranks)
```

Comment:
When we divide by `size(localQ, 1) * size(localQ, 3) * nranks` we are assuming that all threads have the same number of elements and points. Is this always the case? 
@lcw @jkozdon @simonbyrne @kpamnany

Reply:
Each rank might have a different number of elements.  So the assumption you are making is not correct.

Reply:
It might be nice to compute this denominator once.